File: android/src/main/java/com/google/android/libraries/privacy/ppn/internal/UptimeDurationTracker.java
// Copyright 2020 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the );
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an  BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package com.google.android.libraries.privacy.ppn.internal;

import java.time.Clock;
import java.time.Duration;
import java.util.ArrayList;
import java.util.List;

/** Utility for tracking the uptime of a particular metric over an interval. */
class UptimeDurationTracker {
  private static final String TAG = "UptimeDurationTracker";
  private List<Duration> uptimeDurations = new ArrayList<>();
  private final UptimeTracker uptimeTracker = new UptimeTracker();
  private final Object lock = new Object();

  /** Sets the uptime metric to be "started". */
  public void start(Clock clock) {
    synchronized (lock) {
      uptimeTracker.start(clock);
    }
  }

  /** Sets the uptime metric to be stopped. */
  public void stop(Clock clock) {
    synchronized (lock) {
      uptimeTracker.stop(clock);
      captureDuration(clock);
    }
  }

  /** Adds the total uptime to the uptime durations */
  private void captureDuration(Clock clock) {
    Duration duration = uptimeTracker.collectDuration(clock);
    if (!duration.isZero()) {
      uptimeDurations.add(duration);
    }
  }

  /** Returns the List of uptimeDurations and resets list's size to zero. */
  public List<Duration> collectDurations(Clock clock) {
    synchronized (lock) {
      captureDuration(clock);
      List<Duration> durations = uptimeDurations;
      uptimeDurations = new ArrayList<>();
      return durations;
    }
  }
}

File: krypton/desktop/windows/uptime_duration_tracker.cc
/*
 * Copyright (C) 2022 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "privacy/net/krypton/desktop/windows/uptime_duration_tracker.h"

#include <memory>
#include <vector>

#include "privacy/net/krypton/desktop/windows/uptime_tracker.h"
#include "privacy/net/krypton/krypton_clock.h"
#include "third_party/absl/time/time.h"

namespace privacy {
namespace krypton {
namespace windows {

UptimeDurationTracker::UptimeDurationTracker(KryptonClock* clock)
    : uptime_tracker_(clock) {}

void UptimeDurationTracker::Start() { uptime_tracker_.Start(); }

void UptimeDurationTracker::Stop() {
  absl::MutexLock l(&mutex_);
  uptime_tracker_.Stop();
  CaptureDuration();
}

std::vector<absl::Duration> UptimeDurationTracker::CollectDurations() {
  absl::MutexLock l(&mutex_);
  CaptureDuration();

  std::vector<absl::Duration> durations(uptime_durations_.begin(),
                                        uptime_durations_.end());
  uptime_durations_.clear();
  return durations;
}

void UptimeDurationTracker::CaptureDuration() {
  absl::Duration duration = uptime_tracker_.CollectDuration();
  if (duration != absl::ZeroDuration()) {
    uptime_durations_.push_back(duration);
  }
}

}  // namespace windows
}  // namespace krypton
}  // namespace privacy
